File: ios/LowerBackStretching/Views/StretchVideoPlayer.swift
import SwiftUI
import AVKit

/// A muted, looping video player for stretch demonstrations.
///
/// Uses `AVPlayer` behind `VideoPlayer`. The video plays
/// automatically, loops forever, and has no audio. The view
/// handles lifecycle correctly: the player is released when the
/// view leaves the hierarchy.
struct StretchVideoPlayer: View {
    let videoUrl: String

    @State private var player: AVPlayer?

    var body: some View {
        VideoPlayer(player: player)
            .onAppear {
                guard let url = URL(string: videoUrl) else { return }
                let p = AVPlayer(url: url)
                p.isMuted = true
                p.play()
                // Loop the video when it reaches the end.
                NotificationCenter.default.addObserver(
                    forName: .AVPlayerItemDidPlayToEndTime,
                    object: p.currentItem,
                    queue: .main
                ) { _ in
                    p.seek(to: .zero)
                    p.play()
                }
                self.player = p
            }
            .onDisappear {
                player?.pause()
                player = nil
            }
    }
}

File: android/app/src/main/java/com/lowerbackstretching/ui/components/StretchVideoPlayer.kt
package com.lowerbackstretching.ui.components

import android.net.Uri
import androidx.annotation.OptIn
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.viewinterop.AndroidView
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.ui.AspectRatioFrameLayout
import androidx.media3.ui.PlayerView

/**
 * A muted, looping video player for stretch demonstrations.
 *
 * Uses ExoPlayer (Media3) behind a [PlayerView]. The video plays
 * automatically, loops forever, and has no audio. The composable
 * handles lifecycle correctly: the player is released when the
 * composable leaves composition.
 *
 * @param videoUrl HTTP(S) URL to a short looping MP4.
 * @param modifier Applied to the wrapping [Box].
 */
@OptIn(UnstableApi::class)
@Composable
fun StretchVideoPlayer(
    videoUrl: String,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current

    val exoPlayer = remember(context) {
        ExoPlayer.Builder(context).build().apply {
            repeatMode = Player.REPEAT_MODE_ONE
            volume = 0f
            playWhenReady = true
            setMediaItem(MediaItem.fromUri(Uri.parse(videoUrl)))
            prepare()
        }
    }

    DisposableEffect(exoPlayer) {
        onDispose {
            exoPlayer.release()
        }
    }

    Box(modifier = modifier) {
        AndroidView(
            factory = {
                PlayerView(context).apply {
                    player = exoPlayer
                    useController = false
                    resizeMode = AspectRatioFrameLayout.RESIZE_MODE_FIT
                }
            },
            modifier = Modifier.fillMaxSize(),
        )
    }
}
